### DISCUSSION TITLE
Issues with BLE Application Development and Google Play Compliance

### DISCUSSION THREAD

**Discussion by iprimo** [innoveit/react-native-ble-manager#1299] [General]
Hi,

I am working on a BLE-based application and found it much easier to bypass the initial configuration requirements using the example provided in your repository. Thank you for making this process more convenient!

However, I encountered a couple of issues that I’d like to highlight:

**Issue #1:**
When trying to upload the code (not directly your example code but a customized version based on it), I encountered the following errors in Google Play (screenshot attached):

`
Duplicate declarations of <uses-permission> element 'android.permission.ACCESS_COARSE_LOCATION' in manifest with different maxSdkVersions.  
Duplicate declarations of <uses-permission> element 'android.permission.ACCESS_FINE_LOCATION' in manifest with different maxSdkVersions.  

`

To temporarily resolve this, I commented out the **addLocationPermissionToManifest** function in the **withBLEAndroidManifest.js** file. While this allowed the app to work, it limits the range of targeted devices, which is not ideal for the final solution.

**Issue #2:**
New app uploads to Google Play require targeting the latest SDK/API. To meet these requirements, I had to execute the following commands and make changes to the app.json file:

`npx expo install expo-build-properties `

`    "plugins": [
      [
        "expo-build-properties",
        {
          "android": {
            "compileSdkVersion": 34,
            "targetSdkVersion": 34,
            "buildToolsVersion": "34.0.0"
          },
          "ios": {
            "deploymentTarget": "13.4"
          }
        }
      ]
    ],
`

These changes ensured compliance with Google Play's policies but required additional setup steps.

I’d greatly appreciate your efforts in addressing the issues mentioned above.

![ble_bug](https://github.com/user-attachments/assets/91e8518c-573a-4642-a842-5a9d7a22c58e)
e.

**Comment by lucaswitch** [comment]
I see a lot of people struggling with the permissions but we know it's not a library issue since users find it hard to deal with using an
other libraries/languages.
I wonder if would be interesting to add those runtime permissions safety checks on dev mode?
Something just like the browsers do when server does not emit the cross origin http headers.

For example:
- User tries to scan for the first time the library. -
- The library checks if  the scan manifest is right and log a warning if not set.

**Reply by iprimo** [reply]
Certainly, it does not seem to be library problem.
It would be excellent if you could add such runtime check and\or suggest some corrections (if needed).

**Reply by iprimo** [reply]
Hi @lucaswitch , 

Regarding the issues above, I am trying to upgrade the example components to the latest versions and everything fails.

I am wondering if the example in the source code could be upgraded to match\compatible to the latest version of expo\react-native.

Thanks

**Comment by Akyna** [comment]
Hi.
I've got the same issue in Expo

plugin config 

```
    [
      'react-native-ble-manager',
      {
        neverForLocation: true,
        bluetoothAlwaysPermission: 'Allow $(PRODUCT_NAME) to connect printer and barcode scanner',
      },
    ],
```

ISSUE
```
Uploading App Bundle app-release.aab to Google Play failed.
Duplicate declarations of <uses-permission> element 'android.permission.ACCESS_COARSE_LOCATION' in manifest with different maxSdkVersions.
```
Cuz we have these two lines 

```
  <uses-permission-sdk-23 android:name="android.permission.ACCESS_COARSE_LOCATION" android:maxSdkVersion="30"/>
  <uses-permission-sdk-23 android:name="android.permission.ACCESS_FINE_LOCATION" android:maxSdkVersion="30"/>
```

But 
i have 

```
  <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION"/>
  <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION"/>
```
From other library

**Reply by simonschwitzky** [reply]
Having the exact same issue, seems like Google Play lets you publish the app regardless.

**Reply by marcosinigaglia** [reply]
Try to manually change the plugin code and check if it's working with these changes:

```
function addLocationPermissionToManifest(androidManifest, neverForLocationSinceSdk31) {
    const existingPermissionNames = new Set();
    if (Array.isArray(androidManifest.manifest['uses-permission'])) {
        for (const item of androidManifest.manifest['uses-permission']) {
            const permissionName = item?.$?.['android:name'];
            if (permissionName) {
                existingPermissionNames.add(permissionName);
            }
        }
    }
    if (Array.isArray(androidManifest.manifest['uses-permission-sdk-23'])) {
        for (const item of androidManifest.manifest['uses-permission-sdk-23']) {
            const permissionName = item?.$?.['android:name'];
            if (permissionName) {
                existingPermissionNames.add(permissionName);
            }
        }
    }
    const optMaxSdkVersion = neverForLocationSinceSdk31
        ? {
            'android:maxSdkVersion': '30',
        }
        : {};
    const permissionsToAdd = [];
    if (!existingPermissionNames.has('android.permission.ACCESS_COARSE_LOCATION')) {
        permissionsToAdd.push({
            $: {
                'android:name': 'android.permission.ACCESS_COARSE_LOCATION',
                ...optMaxSdkVersion,
            },
        });
    }
    if (!existingPermissionNames.has('android.permission.ACCESS_FINE_LOCATION')) {
        permissionsToAdd.push({
            $: {
                'android:name': 'android.permission.ACCESS_FINE_LOCATION',
                ...optMaxSdkVersion,
            },
        });
    }
    if (permissionsToAdd.length) {
        if (!Array.isArray(androidManifest.manifest['uses-permission-sdk-23'])) {
            androidManifest.manifest['uses-permission-sdk-23'] = [];
        }
        androidManifest.manifest['uses-permission-sdk-23'].push(...permissionsToAdd);
    }
    return androidManifest;
}
```

**Reply by simonschwitzky** [reply]
I tested this patch with a new release and can confirm that it fixes the issue of duplicate permissions getting added to the manifest. Would be awesome to see this land in the package! Thanks a lot.
